iowrappers/photos_client.go

```diff
@@ -123,25 +125,66 @@ func dfs(node *html.Node, judger func(*html.Node) bool, validator func(string) b
 	return "", false
 }
 
-func (photoClient *MapsPhotoClient) GetPhotoURL(ctx context.Context, photoRef string) (PhotoURL, error) {
+func (c *MapsPhotoClient) GetPhotoURL(ctx context.Context, photoRef string, placeId string) (PhotoURL, error) {
 	var photoURL PhotoURL
-	resp, err := photoClient.mapsClient.PlacePhoto(ctx, &maps.PlacePhotoRequest{
+	resp, err := c.mapsClient.client.PlacePhoto(ctx, &maps.PlacePhotoRequest{
 		PhotoReference: photoRef,
 		MaxWidth:       400,
 	})
 	if err != nil {
 		return photoURL, err
 	}
 
-	image, err := resp.Image()
-	if err != nil {
-		return photoURL, err
+	Logger.Debugf("photo response content type is: %s", resp.ContentType)
+	var img image.Image
+	if resp.ContentType == "image/png" {
+		img, err = png.Decode(resp.Data)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		img, err = resp.Image()
+		// content type maybe text/html, need to update place photo reference
+		if err != nil {
+			var r maps.PlaceDetailsResult
```

Comment:
I see there are quite a few new steps added here for error handling starting from line 148, such as refreshing place details, updating redis photo reference and trying to decode the image again. You may want to consider refactoring code into functions.

Reply:
Good point, done.